docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -0,0 +1,261 @@
+# ADR 017: Subspace fee grant
+
+## Changelog
+
+- Nov 14th, 2022: First draft;
+- Nov 21th, 2022: First review.
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+This ADR introduces a new subspace-based fee grant method, which allows subspace owners/admins to pay the fees of subspace-related transactions on behalf of the users.
+
+## Context
+
+One of the major problems of current Web3-based applications that make them complicated to use by common people is the fact that they require users to have some tokens in order to pay transaction fees to perform various operations. For instance, in order to create a post on any Desmos-based social network, users have to:
+1. understand what a _transaction_ is and why they need some DSM to broadcast it;
+2. get some DSM either via an on-ramp service or by swapping existing funds.
+
+The `x/feegrant` module of the Cosmos SDK allows anyone to pay fees on behalf of other users, meaning that the latter can ideally perform any kind of transaction without even knowing the concept of fees. However, the `x/feegrant` allowance is not subspace-specified: this means that subspace admins/owners might unexpected pay the fees for transactions that are related to other subspaces.
+
+## Decision
+
+We will implement a subspace-specified fee grant process based on `x/feegrant` that allows subspace fees providers to pay fees for the users inside the specified subspace. This system will allow subspace owners/admins to issue a fee grant towards either a users group or a single user, so that they can later leverage this grant to execute subspace-related transactions without having to worry about paying fees.
+
+### DeductFeeDecorator
+
+Currently, `x/auth` provides a `DeductFeeDecorator` based on `x/feegrant` to execute the action deducting the fees of a transaction from the signer/feepayer. We will build a new subspace-specified `DeductFeeDecorator` to replace the current one.
+
+The new subspace-specified `DeductFeeDecorator` will work as follows:
+1. check if all the messages in the transaction are messages related to the same subspace;
+2. if the messages in the transaction are all realted to the same subspace and a fee grant exists for that subspace, run the `x/subspaces` `DeductFeeDecorator`; otherwise run the `x/auth` `DeductFeeDecorator`,
+
+### Types
+
+#### Allowance
+
+The `x/feegrant` module already provides an interface (`FeeAllowanceI`) to represent a generic allowance, along with a set of useful implementations like `BasicAllowance`, `AllowedMsgAllowance` and `PeriodicAllowance`. Since we are not going to introduce new kinds of allowances, nor we are going to edit how an allowance is represented, we will reuse these types.
```

Comment:
I was thinking about one thing: should we creae a new implementation of `FeeAllowanceI` named `SubspaceFeeAllowance` that is related to a single subspace? This way, we could even leverage the default `x/feegrant` `MsgGrantFeeAllowance` message to create such allowances.

Reply:
Using the existing ones is enough from my side. 
These cases would be covered:
1. granter grants all the subspace messages with a spend limit by `BasicAllowance`
2. granter grants some of subspace messages with a spend limit by `AllowedMsgAllowance`
3. granter grant all the subspaces messages with periodic limit by `PeriodicAllowance` with `BasicAllowance`
4. granter grant all some of subspaces messages with periodic limit by `PeriodicAllowance` with `AllowedMsgAllowance`

Reply:
@dadamu I see. my idea was to remove the need of handling user-specific grants inside the `x/subspace` module, and instead implement a new `FeeAllowanceI` that can be properly handled by the `x/feegrant` module. Maybe the same could be done for the group as well? This was we don't have to store anything using custom keys but we can simply reply on the `x/feegant` module to store everything. But I'm not sure whether that's possible

Reply:
@RiccardoM 
The subspace group grant is impossible since `feegrant` only gives the possibility to user grant like [this](https://github.com/cosmos/cosmos-sdk/blob/main/x/feegrant/keeper/keeper.go#L43), grantee must be a `sdk.AccAddress`.
On the other hand, if we move subspace user grant into `x/feegrant`, the check would be no longer subspaces-specific, in addition, it would conflict with the non-subspace feegrant.

Reply:
@dadamu Ok, perfect. We can leave it like you designed it then